### ISSUE TITLE
Reloading Tor on Tails at the start of X session could fail in some configurations

### ISSUE THREAD

**Issue by ageis** [freedomofpress/securedrop#982]
As part of PR #955 we moved the Tor reload and the appending of torrc_additions → torrc to the start of the X session (i.e. immediately after login) instead of requiring the user to click on a launcher. Today on one machine running Tails 1.3.2 we found a bug where Tor will not reload successfully at this stage, with the following error:

```
[warn] Failed to parse/validate config: While Tor is running, changing Sandbox is not allowed
```
This is because Tails runs Tor with some local settings including "Sandbox 1", which utilizes the seccomp facility of the Linux kernel. But changing the Sandbox setting is not allowed without a full restart. The problem with reloading at this stage is that Tails is actually appending the "Sandbox 1" line and making torrc modifications some time later. Per the Tails source code, the sandboxing is added if you select the default option not to start Tor Launcher in Tails Greeter, and then Tor is started by a NetworkManager hook when a network link is detected. (see `/etc/NetworkManager/dispatcher.d/10-tor.sh`)
 
```
# The Tor syscall sandbox is not compatible with managed proxies.
# We could possibly detect whether the user has configured any such
# thing via Tor Launcher later (e.g. in 60-tor-ready.sh),
# but then we would have to restart Tor again to enable the sandbox.
# Let's avoid doing that, and enable the Sandbox only if no special Tor
# configuration is needed. Too bad users who simply need to configure
# a HTTP proxy or allowed firewall ports won't get the sandboxing, but
# much better than nothing.
if [ "$(tails_netconf)" = "direct" ]; then
   tor_set_in_torrc Sandbox 1
fi
```
While the seccomp sandbox was added with Tails 1.3, this particular issue would have a higher chance of presenting on slower machines and connections which is why I never observed it. My first idea to deal with this was to add a "sleep" command for a few minutes to the .xsessionrc before we run securedrop_init, but that's contingent on there being a working ethernet connection or WiFi access point saved in persistence that will be ready in the right amount of time. We need a solution that won't allow any race conditions and will work across a variety of hardware. I figured that we might have a loop that waits for Tails to add the "Sandbox 1" before running our Tor reload. But the solution also needs to survive future changes to the software.

I have engaged the Tails developers about this issue and they have encouraged me to e-mail them in more detail. Although the person I spoke to thinks that our use case may be difficult to mix with Tails, they also note that the Tor integration and startup in Tails will change significantly when they move to Jessie, so we should be aware of their changes. Ultimately by digging through the source code and communicating with the Tails developers I think we can find a good way to integrate our need for persistent and automatic hidden service authentication initialization, and I'll be pursuing this.

**Comment by ageis** [comment]
The suggestion I got from one of the Tails developers is to use a NetworkManager hook instead:

/etc/NetworkManager/dispatcher.d/xx-hook_name, where xx < 10

Add this to persistence.conf: /etc/NetworkManager/dispatcher.d source=custom-nm-hooks,link

and then put the hook in /live/persistence/TailsData_unlocked/custom-nm-hooks

We can define our own persistent directories in persistence.conf to do basically anything we want. I'll end up using this for adding custom XDG menus too.

**Comment by ageis** [comment]
Fix implemented [in this commit](https://github.com/ageis/securedrop/commit/b2b022689f72c9f46008ec3768c77d83db8bc262) testing now.

**Comment by ageis** [comment]
@garrettr since we merged the xsessionrc stuff from #925 in develop and 0.3.1 and now uncovered this bug/edge case then if you like I can send the switch to NetworkManager hook as a separate PR/fix so that it doesn't have to be merged with the rest of #955

**Comment by garrettr** [comment]
@ageis Let's keep in in #955, although at this point it's looking like it might be a good idea to rebase that beast.
